src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -1385,36 +1529,70 @@ int POSIXObject::close()
   return 0;
 }
 
-int POSIXObject::read(int64_t ofs, int64_t end, bufferlist& bl,
+int POSIXObject::read(int64_t ofs, int64_t left, bufferlist& bl,
 		      const DoutPrefixProvider* dpp, optional_yield y)
 {
-  int64_t len = std::min(end - ofs + 1, READ_SIZE);
-  ssize_t ret;
+  if (!shadow) {
+    // Normal file, just read it
+    int64_t len = std::min(left + 1, READ_SIZE);
+    ssize_t ret;
 
-  ret = lseek(obj_fd, ofs, SEEK_SET);
-  if (ret < 0) {
-    ret = errno;
-    ldpp_dout(dpp, 0) << "ERROR: could not seek object " << get_name() << " to "
-      << ofs << " :" << cpp_strerror(ret) << dendl;
-    return -ret;
+    ret = lseek(obj_fd, ofs, SEEK_SET);
+    if (ret < 0) {
+      ret = errno;
+      ldpp_dout(dpp, 0) << "ERROR: could not seek object " << get_name() << " to "
+	<< ofs << " :" << cpp_strerror(ret) << dendl;
+      return -ret;
+    }
+
+    ret = ::read(obj_fd, read_buf, len);
+    if (ret < 0) {
+      ret = errno;
+      ldpp_dout(dpp, 0) << "ERROR: could not read object " << get_name() << ": "
+	<< cpp_strerror(ret) << dendl;
+      return -ret;
+    }
+
+    bl.append(read_buf, ret);
+
+    return ret;
   }
 
-  ret = ::read(obj_fd, read_buf, len);
-  if (ret < 0) {
-    ret = errno;
-    ldpp_dout(dpp, 0) << "ERROR: could not read object " << get_name() << ": "
-                  << cpp_strerror(ret) << dendl;
-    return -ret;
+  // It's a multipart object, find the correct file, open it, and read it
+  std::string pname;
+  for (auto part : parts) {
+    if (ofs < part.second) {
+      pname = part.first;
+      break;
+    }
+
+    ofs -= part.second;
   }
 
-  bl.append(read_buf, ret);
+  if (pname.empty()) {
+    // ofs is past the end
+    return 0;
+  }
 
-  return ret;
+  POSIXObject* shadow_obj;
+  std::unique_ptr<rgw::sal::Object> obj = shadow->get_object(rgw_obj_key(pname));
+  shadow_obj = dynamic_cast<POSIXObject*>(obj.get());
```

Comment:
if this can't fail, can we use static_cast?

Reply:
(or even reinterpret_cast if the compiler is annoying)

Reply:
Sure.  I only used dynamic_cast because people have complained when I used static_cast in the past.